### DISCUSSION TITLE
no_std support for bare-metal, wasm, and other runtime-impoverished environments

### DISCUSSION THREAD

**Discussion by emberian** [celo-org/optics-monorepo#637] [Ideas]
we should audit our usage of `std` (and especially needless use of heap collections) and make the core libraries independent of any system runtime. i haven't even started looking at this yet, i just like doing it to every crate i can.

as a bonus, it will make the execution-time footprint lighter and reduce the amount of memory hierarchy touches in the critical paths. although my understanding is that this infrastructure is nowhere near perf-critical.

**Comment by prestwich** [comment]
this is a nice to have, but I am not sure we should spend time on it soon. We expect to have extreme resource surplus for all agents so far
